### PR TITLE
Adds the ability to bundle adjust CSM images

### LINKED ISSUE #4410: Bundle Adjusting CSM images
## Description

With #4291 we now have CSM camera models working in ISIS. The next step is to add the ability to bundle adjust CSM camera models with jigsaw. We've done some planning and have a good idea of how to handle the differences between CSM and ISIS sensors, but there are a few places we would like more input from users.

### Background

First of all, a brief overview of bundle adjustment in CSM vs ISIS. ISIS bundle adjustments works by adjusting constant offsets or polynomial offsets to the position and pointing where as CSM bundle adjustment just adjusts generic parameters. The way that CSM handles bundle adjustment is just a more general way than the way that ISIS does. So, we're planning to leverage this in the actual implementation.

When you adjust a CSM sensor model, you specify a set of parameters that you want to adjust. The parameters available are specific to that model and are completely opaque to the user outside of a name and a [type](https://github.com/sminster/csm/blob/master/csm.h#L90). For example with our USGSCSM line scan model the options are:

* IT Pos. Bias
* CT Pos. Bias
* Rad Pos. Bias
* IT Vel. Bias
* CT Vel. Bias
* Rad Vel. Bias
* Omega Bias
* Phi Bias
* Kappa Bias
* Omega Rate
* Phi Rate
* Kappa Rate
* Omega Accl
* Phi Accl
* Kappa Accl
* Focal Bias

Each parameter can be turned on or off independently and the sensor model will handle the a priori values and covariances between them. You can also specify a [set](https://github.com/sminster/csm/blob/master/csm.h#L111) of parameters you want to work with. There are also more complex options such as parameter sharing that we will likely not get to.

The key difference between ISIS and CSM parameters is that with ISIS we have full knowledge of each parameter and how it translates to the orientation of the model, but with CSM we only know a small amount. For this reason, setting parameters for CSM sensor models is going to require users to be knowledgeable about the parameters the model they are using provides.

### Changes to arguments

This feature should be purely additive and not interfere with how jigsaw currently works. For this reason we will begin by adding a new parameter that specifies a type or set of CSM parameters to adjust. All parameters of the specified type or set will be enabled for all CSM images when used.

For more detailed use cases, we want to extend the current scconfig parameter that allows setting solve parameters on a per sensor basis. We hope to add the ability to specify an exact list of parameter names that you want enabled for a given sensor. For example, if you are a using USGSCSM to adjust Phobos images captured by Viking NAC and HRSC you would have 1 section for the framing sensor and 1 section for the line scan sensor. The frame section would be you solve setting for Viking and the line scan section would be your solve setting for HRSC.

### Changes to output files

With the greater variability in the parameters, we have to have much greater flexibility in the output files. The bundleout.txt file will be a simple modifications where each image will only lists the parameters it has enabled. For the images CSV, though, we will need to make significant changes.

First, we plan to output multiple images CSVs when multiple solve settings are used. Each different solve setting will have its own CSV file. In our previous Phobos examples we would output two CSVs, one for viking images and one for HRSC images. This will allow them to retain their flat file format but allow us to properly handle the differences when adjusting images from multiple sensors.

We also plan to add a separate CSV output for the target body parameters. This CSV will only be a single entry, but it will make it dramatically easier to programmatically read in the target body parameters (currently they have to parsed from the bundleout.txt file).

### Conflicts

The largest conflict is with the target body solution options. Because the CSM API does not surface the rotation from J2000 to the target, we are unable to compute the target rotation. Similarly, we cannot compute the partial derivatives with respect to the target ellipsoid. This means that we will be unable to solve for target body parameters when using ANY CSM sensor models. It is possible to implement this in a specific implementation of the CSM API and we hope to add this to USGSCSM, but we cannot handle this issue on the ISIS side of the API.

IPCE is also a sticking point. We do not have the resources to expand the IPCE interface to allow for CSM adjustments. So, we will maintain the current functionality that allows for adjustments of ISIS sensor models in IPCE, but it will error if you attempt to use a CSM sensor model.

## Example  

In this example we will be adjusting images of Phobos. We will have some number of HRSC, SRC, CTX, and Viking NAC images. The CTX images and Viking NAC images will use USGSCSM provided CSM sensor models and the HRSC and SRC images will use ISIS sensor models.

```
jigsaw fromlist=phobos_cubes.lis cnet=phobos_combined_net.lis scconfig=phobos_solve_params.pvl file_prefix= phobos_combined_solve 
```

The contents of phobos_solve_params.pvl are

```
Object = SensorParameters
# ISIS sensor model parameters
  Group = SRC/MEX_HRSC_SRC
    CamSolve=Angles
    Twist=yes
    Camera_Angles_Sigma=2.0
  EndGroup
  Group = HRSC/MEX_HRSC_S1
    CamSolve=Accelerations
    Twist = Yes
    OverExisting = Yes
    SPSolve = Accelerations
    OverHermite = Yes
    Spacecraft_position_sigma=100.0
    Spacecraft_velocity_sigma=1.0
    Spacecraft_acceleration_sigma=0.1
    Camera_Angles_Sigma=2.0
    Camera_AngularVelocity_Sigma=0.1
    Camera_AngularAcceleration_Sigma=0.01
  EndGroup

# CSM sensor model parameters
  Group = USGS_ASTRO_FRAME_SENSOR_MODEL
    w
    v1
    v2
    v3
  EndGroup
  Group = USGS_ASTRO_LINE_SCANNER_SENSOR_MODEL
    "IT Pos. Bias"
    "CT Pos. Bias"
    "Rad Pos. Bias"
    "Omega Bias"
    "Phi Bias"
    "Kappa Bias"
    "Omega Rate"
    "Phi Rate"
    "Kappa Rate"
  EndGroup
EndObject
End
```

This will solve for just angles on the SRC images, positional acceleration + angular acceleration on HRSC, just angles on Viking, and positions +  angular velocity on CTX. This will produce 4 images CSVs: one for SRC, one for HRSC, one for USGS_ASTRO_FRAME_SENSOR_MODEL and one for USGS_ASTRO_LINE_SCANNER_SENSOR_MODEL

### PR DESCRIPTION
**Pull request by jessemapel**
## Description

This PR merges in the csmbundle feature branch which adds the ability to bundle adjust CSM images (images that have gone through csminit and use a CSM camera model). This is primarily a change to jigsaw application and the Bundle* classes. These changes ARE API BREAKING as we have had to modify the API of several classes, primarily the BundleObservation class.

Due to the more variable nature of CSM model parameters, this also changes how jigsaw's images.csv is output. Each different sensor has its own images.csv file. So if you bundle adjust some LRO NAC pairs and Apollo Metric images, you would get 3 images.csv files:

1. LRO NAC Left
2. LRO NAC Right
3. Apollo Metric

This allows for use to actually have true headers in the images.csv as opposed to the mangled headers we get currently when doing per-sensor solves with jigsaw's scconfig option.

This PR also handles a number of other jigsaw related things:

- Removed the SIP wrappers as the API changes break them and they are no longer supported
- Removed a deprecated bundle output function from #3404
- Moved the last of the jigsaw application tests to the new Functional gtest format

We also checked several control network generation and editing applications for compatibility with CSM. We did not encounter issues with any of the following applications:

- footprintinit
- findimageoverlaps
- seedgrid
- autoseed
- cnetadd
- pointreg
- qnet

## Related Issue

Fixes #4410 

## Motivation and Context

This is a major improvement in CSM functionality in ISIS. Now, users can perform bundle adjustments in ISIS and then export the updated CSM state strings to other SETs for further processing.

## How Has This Been Tested?

We hand tested everything with a demo CTX network using USGSCSM. All new functions are fully unitTested using a new Mock CSM fixture and jigsaw has had a CSM specific test that bundle adjusts a fully synthetic data set.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/control/objs/BundleUtilities/BundleObservationVector.cpp [hunk 1]
```diff
@@ -109,27 +117,36 @@ namespace Isis {
 
       bundleImage->setParentObservation(bundleObservation);
 
-      // update observation number to observation ptr map
-      m_observationNumberToObservationMap.insertMulti(observationNumber,bundleObservation);
+      // updateo observation number to observation ptr map
```
Comment (kaitlyndlee) [inline isis/src/control/objs/BundleUtilities/BundleObservationVector.cpp]: Small typo: "updateo" -> "update"

File: isis/src/control/objs/CsmBundleObservation/CsmBundleObservation.cpp [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+
+  /**
+   * Constructs a CsmBundleObservation initialized to a default state.
+   */
+  CsmBundleObservation::CsmBundleObservation() {
+  }
+
+
+  /**
+   * Constructs a CsmBundleObservation from an BundleImage, an instrument id, an observation
```
Comment (kaitlyndlee) [inline isis/src/control/objs/CsmBundleObservation/CsmBundleObservation.cpp]: Typo: "an BundeImage" -> "a BundleImage"

**Review by kaitlyndlee — COMMENTED** [review]
Just those two typos, otherwise looks good!

**Comment by jessemapel** [comment]
@kaitlyndlee Fixed the typos

**Review by kaitlyndlee — APPROVED** [review]

**Pull request merged**